src/Toml/Parser/Value.hs

```diff
@@ -41,14 +59,47 @@ decimalP = zero <|> more
     check :: Maybe Integer -> Parser Integer
     check = maybe (fail "Not an integer") pure
 
+-- | Parser for hexadecimal numbers : included parsing of underscore.
+hexadecimalP :: Parser Integer
+hexadecimalP = zero <|> more
```

Comment:
All 3 parsers have a similar structure. I propose to remove code duplication by introducing a separate function that takes a parser for digit, string tag and a conversion function. If `zero` is redundant in `decimal` as well, then we might implement all 4 functions using the same common parsing function.